olive/passes/pytorch/qlora.py

```diff
@@ -0,0 +1,487 @@
+                param.data = param.data.to(compute_dtype)
+
+        # add lora modules
+        logger.debug("Adding LoRA modules")
+        # this doesn't pick up the embedding layer and projection layer since those are not quantized
+        # this is good since we don't want to touch those, LoRA might not work with input output embedding layers
+        modules = cls.find_all_linear_names(pytorch_model)
+        lora_config = LoraConfig(
+            r=config["lora_r"],
+            lora_alpha=config["lora_alpha"],
+            lora_dropout=config["lora_dropout"],
+            target_modules=modules,
+            bias="none",
+            task_type=peft_task_type,
+        )
+        pytorch_model = get_peft_model(pytorch_model, lora_config)
+
+        # cast lora modules to compute_dtype
+        for module in pytorch_model.modules():
+            if isinstance(module, LoraLayer):
+                # TODO: why only cast for bfloat16? https://github.com/artidoro/qlora/blob/main/qlora.py#L397
+                module.to(compute_dtype)
+
+        return pytorch_model, model_loading_args, tokenizer
+
+    @staticmethod
+    def smart_tokenizer_and_embedding_resize(
+        special_tokens_dict: Dict, tokenizer: transformers.PreTrainedTokenizer, model: transformers.PreTrainedModel
+    ):
+        """
+        Resize the tokenizer and the model embedding layer to take into account new special tokens.
+        """
+        # resize tokenizer
+        num_new_tokens = tokenizer.add_special_tokens(special_tokens_dict)
+        # resize model embedding layer
+        model.resize_token_embeddings(len(tokenizer))
+        if num_new_tokens > 0:
+            logger.info(f"Added {num_new_tokens} new tokens to tokenizer and resized model embedding layer.")
+            input_embeddings_data = model.get_input_embeddings().weight.data
+            output_embeddings_data = model.get_output_embeddings().weight.data
+
+            # average the embeddings of the pre-existing tokens
+            input_embeddings_avg = input_embeddings_data[:-num_new_tokens].mean(dim=0, keepdim=True)
+            output_embeddings_avg = output_embeddings_data[:-num_new_tokens].mean(dim=0, keepdim=True)
+
+            # set the new embeddings to the average
+            input_embeddings_data[-num_new_tokens:] = input_embeddings_avg
+            output_embeddings_data[-num_new_tokens:] = output_embeddings_avg
+
+    @staticmethod
+    def find_all_linear_names(model: torch.nn.Module) -> List[str]:
+        """
+        Find all linear layers in a model.
+        """
+        import bitsandbytes as bnb
+
+        linear_cls = bnb.nn.Linear4bit
+        lora_module_names = set()
+        for name, module in model.named_modules():
+            if isinstance(module, linear_cls):
+                lora_module_names.add(name.split(".")[-1])
+        return list(lora_module_names)
+
+    @staticmethod
+    def get_datasets(config: Dict[str, Any], data_root: str) -> tuple:
```

Comment:
>get_datasets [](http://example.com/codeflow?start=8&length=12)

do we need convert the config dict to the config_class before call any validate_config? 

In this way, all the validate_config could be safely removed?

Reply:
That's a great idea. Done!